### ISSUE TITLE
Pygame interface doesn't work without scaling=native

### ISSUE THREAD

**Issue by Danitegue** [robhagemans/pcbasic#53]
The default option for --scaling [smooth] makes that pcbasic cannot be launch with pygame interface. (It always breaks in pcbasic\interface\video_pygame, line 114).

        try:
            if not self.fullscreen:
                pygame.display.set_mode(self.display_size, 0)
            self._resize_display(*self.display_size)
        except pygame.error:
            self._close_pygame()
            logging.warning('Could not initialise PyGame display')
            raise base.InitFailed()

This is solved by setting in the PCBASIC.INI the option scaling=native.

So, I was wondering if the best solution for having the maximum compatibility is to set the default option to native, instead of smooth, or to have a workaround in this code to switch to native if smooth fails.

Regards

**Comment by robhagemans** [comment]
Hi, some kind of fallback is probably a good idea.

However, before creating workarounds, I'd like to understand a bit more about this issue, since it doesn't happen on my setup:
- does it work if you use `scaling=crisp` ?
- does it work with smooth scaling if you fix the display size (e.g. `dimensions=640,480`)
- does it work if you use `interface=sdl2`?

Finally, could you maybe log the Pygame error like so (and I should change this in the code anyway):
```
except pygame.error as e:
        self._close_pygame()
        logging.warning('Could not initialise PyGame display: %s', e)
        raise base.InitFailed()
```

**Comment by Danitegue** [comment]
With both scaling=crisp or scaling=smooth, the program breaks exactly in the same line of pcbasic\interface\video_pygame.py:

```
try:
        if not self.fullscreen:
            pygame.display.set_mode(self.display_size, 0) -> THE TRY STATMENT BREAKS HERE
        self._resize_display(*self.display_size)
    except pygame.error:
        self._close_pygame() -> THE PROGRAM FINISH HERE: Process finished with exit code -1073740940 (0xC0000374)
        logging.warning('Could not initialise PyGame display') -> SO, THIS LINE IS NEVER EXECUTED. 
        raise base.InitFailed()
```

With scaling=smooth + dimensions=640,480 it works fine.

With scaling=smooth + interface=sdl2 (after downloading the sdl2.dll file into the \pcbasic\pcbasic\interface folder) it works fine.

I would say at least to move the logging.warning line before the self._close_pygame() line, in order to write log the error before the crash, and add a second try to force the smooth + 640,480 if the smooth only fails.

**Comment by robhagemans** [comment]
What's your actual display size? Do you have your screen taller than wide?

I'm working on a hypothesis that the issue is caused by the way the window size is determined, which could be fixable. It would be useful to know what error message pygame throws (i.e.the output from the modified logging code above)
Thanks!

**Comment by Danitegue** [comment]
Oh, sorry I didn't noticed your modified logging code.

So, running with the pygame interface, scaling=smooth, moving the logging line before the self._close_pygame(), and with your modified logging line:

```
        try:
            if not self.fullscreen:
                pygame.display.set_mode(self.display_size, 0)
            self._resize_display(*self.display_size)
        except pygame.error as e:
            logging.warning('Could not initialise PyGame display: %s', e)
            self._close_pygame()
            raise base.InitFailed()

```
We obtain:
WARNING: Could not initialise PyGame display: DDraw created surface with wrong pitch

**Comment by Danitegue** [comment]
I've tested it in two different laptops, one is using 1920x1080 and the other 1366x768. The same crash in both.

**Comment by Danitegue** [comment]
In the case of the 1920x1080 laptop, 

- when trying to run with scaling=native,  the (NON) breaking line pygame.display.set_mode(self.display_size, 0) is trying to set the pygame display to 672x504. -> No problem. (Proportion x/y = 672/504 = 1.3333... )

- when trying to run with scaling=smooth, the breaking line pygame.display.set_mode(self.display_size, 0) is trying to set the pygame display to 1285x963. -> It crashes. (Proportion x/y = 1285/963= 1.33437...)

I did the following test: In the second case, re-calculate the x as y*1.333... = 963*1.333... = 1284 -> Test with smooth but changing manually the resolution to 1284x963 in the breaking line -> No crash.

So yes, seems that for making it work, the proportion between x/y should be always 4/3=1.333...

**Comment by robhagemans** [comment]
Ok, thanks. That seems a bit more complicated than I thought. Which version of Pygame are you using?

Also, when you say it crashes, what happens exactly? I would expect it simply to close the graphical window and fall back on a command line interface. Is there a Python exception? If so, does it print an error message?

**Comment by Danitegue** [comment]
I'm using pygame 1.9.3. No python exception at all, I cannot debug beyond that point (Process finished with exit code -1073740940 (0xC0000374)).

I have traced the problem until this function:

pcbasic\pcbasic\interface\video_graphical.py, function def _find_display_size()

```
    def _find_display_size(self, canvas_x, canvas_y, border_width):
        """Determine the optimal size for the display."""
        # comply with requested size unless we're fullscreening
        if self.force_display_size and not self.fullscreen:
            return self.force_display_size
        if not self.force_native_pixel:
            # this assumes actual display aspect ratio is wider than 4:3
            # scale y to fit screen
            canvas_y = (1 - display_slack/100.) * (
                        self.physical_size[1] // int(1 + border_width/100.))
            # scale x to match aspect ratio
            canvas_x = (canvas_y * self.aspect[0]) / self.aspect[1]
            # add back border
            pixel_x = int(canvas_x * (1 + border_width/100.)) -> It is assigned 1285 in stead of 1284.
            pixel_y = int(canvas_y * (1 + border_width/100.)) -> It is assigned 963
            return pixel_x, pixel_y
```

The length of the borders can make that the final results of pixel_x/pixel_y does not agree with the aspect ratio.

So, I would suggest to add these few lines to the end, in order to recalculate the pixel_x value in order to maintain the aspect ratio after the borders calculations:

```
    def _find_display_size(self, canvas_x, canvas_y, border_width):
        """Determine the optimal size for the display."""
        # comply with requested size unless we're fullscreening
        if self.force_display_size and not self.fullscreen:
            return self.force_display_size
        if not self.force_native_pixel:
            # this assumes actual display aspect ratio is wider than 4:3
            # scale y to fit screen
            canvas_y = (1 - display_slack/100.) * (
                        self.physical_size[1] // int(1 + border_width/100.))
            # scale x to match aspect ratio
            canvas_x = (canvas_y * self.aspect[0]) / self.aspect[1]
            # add back border
            pixel_x = int(canvas_x * (1 + border_width/100.))
            pixel_y = int(canvas_y * (1 + border_width/100.))
            ratio=float(self.aspect[0]) / float(self.aspect[1])
            if float(pixel_x)/float(pixel_y) != ratio:
                #Recalculate the pixel_x if the pixel_x/pixel_y ratio does not agree with the aspect ratio
                pixel_x = int(pixel_y * (float(self.aspect[0]) / float(self.aspect[1])))
            return pixel_x, pixel_y
```

Whit this, the problem is solved.

**Comment by robhagemans** [comment]
Hi, thanks for the info. I need to do some research on this - it may be a bug in Pygame 1.9.3 on Windows (I've been using 1.9.2 there so far and 1.9.3 only on Linux.).  I just don't see why Pygame would be unable to create a window with a given pixel size. 

I don't think fixing the aspect ratio is the right thing to do, as it will give the screen an incorrect aspect ratio for some video modes and some choices of border. But most importantly, I don't like adding a quick fix for a problem I don't understand, as it'll come back to bite me later on...

Is there a reason you want to use Pygame, by the way? I'm asking because SDL2 is the preferred and standard graphics library for this version of PC-BASIC and I was planning to remove Pygame support in favour of SDL2 in future . Mainly because it's a big, quite old and sparsely maintained library that does many things I don't need, while apparently causing a few hard to debug issues like this one.

**Comment by Danitegue** [comment]
Ok I understand. I really dont know why pygame is not able to run without
this fixed aspect ratio.

I started using pygame because I could not make pcbasic work with other
option while debugging with pycharm. Today I realized about the needed
sdl2.dll file to make the sdl interface work. So no problem if you delete
the support for pygame, I can continue with sdl.

It's enough for today, regards from Tenerife and happy new year!

**Comment by robhagemans** [comment]
I traced this issue back to a problem with environment variables, which I fixed by 013417cd.

I think this solves the issue so I'm closing the ticket. If the problem does come back, feel free to re-open or open a new ticket. Thanks for helping to find and debug this issue!
